third_party/move/move-prover/boogie-backend/src/spec_translator.rs

```diff
@@ -1147,18 +1147,34 @@ impl<'env> SpecTranslator<'env> {
         mid.qualified_inst(sid, inst.to_owned())
     }
 
+    fn get_memory_inst_from_node_opt(&self, node_id: NodeId) -> Option<QualifiedInstId<StructId>> {
+        let mem_ty = &self.get_node_instantiation(node_id)[0];
+        if let Some((mid, sid, inst)) = mem_ty.require_struct_opt() {
+            Some(mid.qualified_inst(sid, inst.to_owned()))
+        } else {
+            None
+        }
+    }
+
     fn translate_resource_exists(
         &self,
         node_id: NodeId,
         args: &[Exp],
         memory_label: &Option<MemoryLabel>,
     ) {
-        let memory = &self.get_memory_inst_from_node(node_id);
-        emit!(
-            self.writer,
-            "$ResourceExists({}, ",
-            boogie_resource_memory_name(self.env, memory, memory_label),
-        );
+        let memory_opt = &self.get_memory_inst_from_node_opt(node_id);
+        let mut resource_name = "".to_string();
+        if let Some(memory) = memory_opt {
+            resource_name = boogie_resource_memory_name(self.env, memory, memory_label);
+        } else {
+            let mem_ty = &self.get_node_instantiation(node_id)[0];
+            if let Type::TypeParameter(arg) = mem_ty {
+                resource_name = format!("#{}_$memory", arg)
```

Comment:
Unfortunately, this wont work like this, because we do not declare any of this `#<typeparam_number>_$memory` variables. I assume your test cases did not hit this case, otherwise the boogie should have failed to compile.

The reason why you did not hit this might be monomorphization. We should try to see whether we can just always report an error also if its a type parameter, and how far we get with this.

Reply:
> Unfortunately, this wont work like this, because we do not declare any of this `#<typeparam_number>_$memory` variables. I assume your test cases did not hit this case, otherwise the boogie should have failed to compile.
> 
> The reason why you did not hit this might be monomorphization. We should try to see whether we can just always report an error also if its a type parameter, and how far we get with this.

I see. Actually, the following code is generated in the boogie file:
```
type #0;
function {:inline} $IsEqual'#0'(x1: #0, x2: #0): bool { x1 == x2 }
function {:inline} $IsValid'#0'(x: #0): bool { true }
var #0_info: $TypeParamInfo;
var #0_$memory: $Memory #0;
```
I will dig more into this

Reply:
Interesting. The boogie code above is the monomorphization of type parameter 0. The type parameter is turned into a given type (`type #0;`). I'm surprised to see we also generate a memory variable for it. I think we may do this because `#0` has the `key` ability. So we could be good. Unfortunately, it is still not clear how to deal with `old(...exists...)`, however.